packages/server/src/sdk/workspace/resources/index.ts

```diff
@@ -262,9 +275,332 @@ function isWorkspaceApp(doc: AnyDocument): doc is WorkspaceApp {
+  sourceProdWorkspaceId: string,
+  destinationProdWorkspaceId: string
+) => {
+  if (!key?.startsWith(`${sourceProdWorkspaceId}/`)) {
+    return key
+  }
+  if (sourceProdWorkspaceId === destinationProdWorkspaceId) {
+    return key
+  }
+  const suffix = key.slice(sourceProdWorkspaceId.length)
+  return `${destinationProdWorkspaceId}${suffix}`
+}
+
+async function copyAttachmentToWorkspace(
+  key: string,
+  destinationKey: string,
+  cache: Map<string, Promise<string | undefined>>
+): Promise<string | undefined> {
+  if (!key || key === destinationKey) {
+    return key
+  }
+  if (!cache.has(key)) {
+    cache.set(
+      key,
+      (async () => {
+        try {
+          const alreadyExists = await objectStore.objectExists(
+            ObjectStoreBuckets.APPS,
+            destinationKey
+          )
+          if (alreadyExists) {
+            return destinationKey
+          }
+        } catch (err) {
+          logging.logWarn(
+            "Resource duplication: failed to check attachment existence",
+            {
+              err,
+              key,
+              destinationKey,
+            }
+          )
+        }
+        try {
+          const { stream, contentType } = await objectStore.getReadStream(
+            ObjectStoreBuckets.APPS,
+            key
+          )
+          await objectStore.streamUpload({
+            bucket: ObjectStoreBuckets.APPS,
+            stream,
+            filename: destinationKey,
+            type: contentType,
+          })
+          return destinationKey
+        } catch (err) {
+          logging.logWarn("Resource duplication: failed to copy attachment", {
+            err,
+            key,
+            destinationKey,
+          })
+          return undefined
+        }
+      })()
+    )
+  }
+  const copiedKey = await cache.get(key)!
+  if (copiedKey === undefined) {
+    cache.delete(key)
+  }
+  return copiedKey
+}
+
+async function remapAttachmentValue(
+  attachment: RowAttachment,
+  context: AttachmentCopyContext
+): Promise<RowAttachment> {
+  if (
+    !attachment?.key ||
+    !context.sourceProdWorkspaceId ||
+    !context.destinationProdWorkspaceId ||
+    context.sourceProdWorkspaceId === context.destinationProdWorkspaceId
+  ) {
+    return attachment
+  }
+
+  const destinationKey = buildDestinationAttachmentKey(
+    attachment.key,
+    context.sourceProdWorkspaceId,
+    context.destinationProdWorkspaceId
+  )
+  if (!destinationKey || destinationKey === attachment.key) {
+    return attachment
+  }
+
+  const copiedKey = await copyAttachmentToWorkspace(
+    attachment.key,
+    destinationKey,
+    context.cache
+  )
+
+  if (!copiedKey) {
+    return attachment
+  }
+
+  return {
+    ...attachment,
+    key: copiedKey,
+    url: "",
+  }
+}
+
+async function remapRowAttachments(
+  row: Row,
+  columns: AttachmentColumn[],
+  context: AttachmentCopyContext
+) {
+  if (!columns.length) {
+    return
+  }
+
+  const rowData = row as Record<string, any>
```

Comment:
NAB: Could maybe add a type instead of **any** ? Even **unknown** is a little more type safe